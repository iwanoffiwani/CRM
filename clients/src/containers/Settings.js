import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Layout from "./Layout";
import { makeStyles } from "@material-ui/core/styles";
import { Paper } from "@material-ui/core";
import { Grid } from "@material-ui/core";
import Typography from "@material-ui/core/Typography";
import EditFields from "./EditFields";
import EditStatuses from "./EditStatuses";

const useStyles = makeStyles(theme => ({
  box: {
    marginBottom: theme.spacing(4)
  },
  paper: {
    padding: "1em"
  },
  title: {
    marginBottom: theme.spacing(2),
    fontWeight: "bold",
    fontSize: "1.5em"
  }
}));

const mapStateToProps = state => {
  return {
    fields: state.fields.payload,
    statuses: state.statuses.payload
  };
};

export const Settings = props => {
  const classes = useStyles();

  return (
    <Layout>
      <div className={classes.root}>
        <Grid container spacing={3}>
          <Grid item lg={4} md={4} sm={12}>
            <Paper className={classes.paper} elevation={3}>
              <Typography className={classes.title} variant="h2" noWrap>
                Поля
              </Typography>
              <EditFields fields={props.fields} />
            </Paper>
          </Grid>
          <Grid item lg={4} md={4} sm={12}>
            <Paper className={classes.paper} elevation={3}>
              <Typography className={classes.title} variant="h2" noWrap>
                Статусы
              </Typography>
              <EditStatuses statuses={props.statuses} />
            </Paper>
          </Grid>
        </Grid>
      </div>
    </Layout>
  );
};

Settings.propTypes = {
  fields: PropTypes.array.isRequired,
  statuses: PropTypes.array.isRequired
};

export default connect(mapStateToProps)(Settings);
