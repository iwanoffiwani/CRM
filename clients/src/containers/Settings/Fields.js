import React, { useState } from 'react'
import { connect } from 'react-redux'
import { makeStyles } from '@material-ui/core/styles'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import Fab from '@material-ui/core/Fab'
import EditIcon from '@material-ui/icons/Edit'
import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline'
import CheckIcon from '@material-ui/icons/Check'
import ClearIcon from '@material-ui/icons/Clear'
import ListItemText from '@material-ui/core/ListItemText'
import TextField from '@material-ui/core/TextField'

const useStyles = makeStyles(theme => ({
  icon: {
    width: 42,
    height: 42,
    marginRight: theme.spacing(2)
  }
}))

const DeleteField = props => {
  const classes = useStyles()
  return (
    <ListItem>
      <Fab className={classes.icon} onClick={() => props.deleteHandler()}>
        <CheckIcon />
      </Fab>
      <Fab className={classes.icon} onClick={() => props.cancelEdit()}>
        <ClearIcon />
      </Fab>
      <ListItemText>
        Вы уверены, что хотите удалить это поле безвозвратно?
      </ListItemText>
    </ListItem>
  )
}

const EditField = props => {
  const classes = useStyles()

  const initialState = { 
    name: props.name 
  }

  const [ state, setState ] = useState(initialState)

  return (
    <ListItem>
      <Fab className={classes.icon} onClick={() => props.editHandler(state.name)}>
        <CheckIcon />
      </Fab>
      <Fab className={classes.icon} onClick={() => props.cancelEdit()}>
        <ClearIcon />
      </Fab>
      <TextField
        value={state.name}
        onChange={e => setState({ ...state, name: e.target.value })}
        placeholder='Введите новое имя'
      />
    </ListItem>
  )
}

const EditFields = props => {
  const classes = useStyles()

  const initialState = {
    query: [
      ...props.fields
    ],
    edit: {
      id: false,
      name: null
    },
    delete: {
      id: false
    }
  }

  const [ state, setState ] = useState(initialState)

  const handlerClickEdit = id => {
    return setState({
      ...state,
      edit: {
        id
      }
    })
  }

  const handlerClickDelete = id => {
    return setState({
      ...state,
      delete: {
        id
      }
    })
  }

  const handlerEditField = name => {
    return setState({
      ...state,
      query: props.fields.map(field => {
        if (field._id === state.edit.id) 
          return {
            ...field,
            name
          }
        
        return field
      })
    })
  }

  const handlerDeleteField = id => {
    return setState({
      ...state,
      query: props.fields.filter(field => 
        field._id !== state.delete.id)
    })
  }

  const handlerCancel = () => {
    return setState({ 
      ...initialState 
    })
  }

  return (
    <List>
      {(() => {
        if (state.edit.id)
          return (
            props.fields.map((field, index) => 
              state.edit.id === field._id ? 
              <EditField 
                key={index}
                name={field.name}
                cancelEdit={handlerCancel}
                editHandler={handlerEditField}
              /> :
              <ListItem key={index} disabled>
                <Fab className={classes.icon}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {field.name}
                </ListItemText>
              </ListItem>
            )
          )
        else if (state.delete.id)
          return (
            props.fields.map((field, index) => 
              state.delete.id === field._id ? 
              <DeleteField 
                key={index}
                cancelEdit={handlerCancel}
                deleteHandler={handlerDeleteField}
              /> :
              <ListItem key={index} disabled>
                <Fab className={classes.icon}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {field.name}
                </ListItemText>
              </ListItem>
            )
          )
        else 
          return (
            props.fields.map((field, index) => 
              <ListItem key={index}>
                <Fab className={classes.icon} onClick={() => handlerClickEdit(field._id)}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon} onClick={() => handlerClickDelete(field._id)}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {field.name}
                </ListItemText>
              </ListItem>
            )
          )
      })()}
    </List>
  )
}

const mapStateToProps = state => {
  return {
    fields: state.fields.payload
  }
}

export default connect(mapStateToProps)(EditFields)