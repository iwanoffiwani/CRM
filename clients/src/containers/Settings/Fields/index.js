import React, { useState } from 'react'
import { connect } from 'react-redux'
import List from '@material-ui/core/List'
import FieldsContext from './FieldsContext'
import FieldItem from './FieldItem'
import EditField from './EditField'
import DeleteField from './DeleteField'
import CreateField from './CreateField'

const SettingFields = props => {

  const initialState = {
    field: {
      edit: {
        id: null
      },
      delete: {
        id: null
      },
      create: {
        init: false
      }
    },
    func: {
      create: () => setState({
        ...state,
        field: {
          ...state.field,
          create: {
            init: true
          }
        }
      }),
      edit: id => setState({ 
        ...state, 
        field: { 
          ...state.field, 
          edit: { 
            id 
          } 
        } 
      }),
      delete: id => setState({
        ...state,
        field: {
          ...state.field,
          delete: {
            id
          }
        }
      }),
      default: () => setState({ ...initialState })
    }
  }

  const [ state, setState ] = useState(initialState)

  return (
    <FieldsContext.Provider value={state}>
      <List>
        {(() => {
          if (state.field.edit.id !== null)
            return (
              props.fields.map((field, index) => 
                state.field.edit.id === field._id ? 
                  <EditField 
                    key={index} 
                    field={field}
                  /> : 
                  <FieldItem 
                    key={index} 
                    field={field}
                  /> 
              )
            )
          else if (state.field.delete.id !== null)
            return (
              props.fields.map((field, index) => 
                state.field.delete.id === field._id ? 
                  <DeleteField 
                    key={index} 
                    field={field}
                  /> : 
                  <FieldItem 
                    key={index} 
                    field={field}  
                  /> 
              )
            )
          else if (!state.field.edit.id && !state.field.delete.id) 
            return (
              props.fields.map((field, index) => 
                <FieldItem 
                  key={index} 
                  field={field}
                /> 
              )
            )
        })()}
        <CreateField />
      </List>
    </FieldsContext.Provider>
  )
}

const mapStateToProps = state => {
  return {
    fields: state.fields.payload
  }
}

export default connect(mapStateToProps)(SettingFields)