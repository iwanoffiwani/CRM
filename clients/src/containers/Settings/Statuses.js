import React, { useState } from 'react'
import { connect } from 'react-redux'
import axios from 'axios'
import { fetchUpdateOrderList, fetchStatuses } from '../../redux/actions'
import { makeStyles } from '@material-ui/core/styles'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import Fab from '@material-ui/core/Fab'
import EditIcon from '@material-ui/icons/Edit'
import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline'
import CheckIcon from '@material-ui/icons/Check'
import ClearIcon from '@material-ui/icons/Clear'
import ListItemText from '@material-ui/core/ListItemText'
import TextField from '@material-ui/core/TextField'
import AddIcon from '@material-ui/icons/Add'

const useStyles = makeStyles(theme => ({
  icon: {
    width: 42,
    height: 42,
    marginRight: theme.spacing(2)
  }
}))

const DeleteStatus = props => {
  const classes = useStyles()

  return (
    <ListItem>
      <Fab className={classes.icon} onClick={() => props.deleteHandler()}>
        <CheckIcon />
      </Fab>
      <Fab className={classes.icon} onClick={() => props.cancelEdit()}>
        <ClearIcon />
      </Fab>
      <ListItemText>
        Вы уверены, что хотите удалить это поле безвозвратно?
      </ListItemText>
    </ListItem>
  )
}

const EditStatus = props => {
  const classes = useStyles()

  const initialState = { 
    name: props.name 
  }

  const [ state, setState ] = useState(initialState)

  return (
    <ListItem>
      <Fab className={classes.icon} onClick={() => props.editHandler(state.name)}>
        <CheckIcon />
      </Fab>
      <Fab className={classes.icon} onClick={() => props.cancelEdit()}>
        <ClearIcon />
      </Fab>
      <TextField
        value={state.name}
        onChange={e => setState({ ...state, name: e.target.value })}
        placeholder='Введите новое имя'
      />
    </ListItem>
  )
}

const AddStatus = props => {
  const classes = useStyles()

  const initialState = { name: '' }

  const [ state, setState ] = useState(initialState)

  return (
    <ListItem>
      <Fab className={classes.icon} onClick={() => props.addStatus(state.name)}>
        <CheckIcon />
      </Fab>
      <Fab className={classes.icon} onClick={() => props.cancelEdit()}>
        <ClearIcon />
      </Fab>
      <TextField
        value={state.name}
        onChange={e => setState({ ...state, name: e.target.value })}
        placeholder='Введите имя поля'
      />
    </ListItem>
  )
}

const EditStatuses = props => {
  const classes = useStyles()

  const initialState = {
    query: [
      ...props.statuses
    ],
    edit: {
      id: false,
      name: null
    },
    delete: {
      id: false
    }
  }

  const [ state, setState ] = useState(initialState)

  const handlerClickEdit = id => {
    return setState({
      ...state,
      edit: {
        id
      }
    })
  }

  const handlerClickDelete = id => {
    return setState({
      ...state,
      delete: {
        id
      }
    })
  }

  const handlerEditStatus = name => {
    return axios({
      method: 'PATCH',
      url: `/api/statuses/`,
      params: {
        id: state.edit.id
      },
      data: {
        name
      }
    })
    .then(() => setState({ 
      ...initialState 
    })) 
    .then(() => props.statusesUpdate())
    .then(() => props.ordersUpdate())
  }

  const handlerDeleteStatus = () => {
    return axios({
      method: 'DELETE',
      url: `/api/statuses/`,
      params: {
        id: state.delete.id
      }
    })
    .then(() => setState({ 
      ...initialState 
    })) 
    .then(() => props.statusesUpdate())
    .then(() => props.ordersUpdate())
  }

  const handlerCancel = () => {
    return setState({ 
      ...initialState 
    })
  }

  const handlerClickAddStatus = () => {
    return setState({
      ...state,
      add: true
    })
  }

  const handlerAddStatus = name => {
    return axios({
      method: 'POST',
      url: `/api/statuses/`,
      data: {
        name
      }
    })
    .then(() => setState({ 
      ...initialState 
    })) 
    .then(() => props.statusesUpdate())
    .then(() => props.ordersUpdate())
  }

  return (
    <List>
      {(() => {
        if (state.edit.id)
          return (
            props.statuses.map((status, index) => 
              state.edit.id === status._id ? 
              <EditStatus 
                key={index}
                name={status.name}
                cancelEdit={handlerCancel}
                editHandler={handlerEditStatus}
              /> :
              <ListItem key={index} disabled>
                <Fab className={classes.icon}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {status.name}
                </ListItemText>
              </ListItem>
            )
          )
        else if (state.delete.id)
          return (
            props.statuses.map((status, index) => 
              state.delete.id === status._id ? 
              <DeleteStatus 
                key={index}
                cancelEdit={handlerCancel}
                deleteHandler={handlerDeleteStatus}
              /> :
              <ListItem key={index} disabled>
                <Fab className={classes.icon}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {status.name}
                </ListItemText>
              </ListItem>
            )
          )
        else 
          return (
            props.statuses.map((status, index) => 
              <ListItem key={index}>
                <Fab className={classes.icon} onClick={() => handlerClickEdit(status._id)}>
                  <EditIcon />
                </Fab>
                <Fab className={classes.icon} onClick={() => handlerClickDelete(status._id)}>
                  <DeleteOutlineIcon />
                </Fab>
                <ListItemText>
                  {status.name}
                </ListItemText>
              </ListItem>
            )
          )
      })()}
      {
        !state.add ? 
        <ListItem>
          <Fab className={classes.icon} onClick={handlerClickAddStatus}>
            <AddIcon />
          </Fab>
          <ListItemText>
            Добавить поле
          </ListItemText>
        </ListItem>
        : <AddStatus 
            addStatus={handlerAddStatus}
            cancelEdit={handlerCancel}
          />
      }
    </List>
  )
}

const mapStateToProps = state => {
  return {
    statuses: state.statuses.payload
  }
}

const mapDispatchToProps = dispatch => {
  return {
    statusesUpdate: () => dispatch(fetchStatuses()),
    ordersUpdate: () => dispatch(fetchUpdateOrderList())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditStatuses)