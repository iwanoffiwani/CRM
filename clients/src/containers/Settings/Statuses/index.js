import React, { useState } from 'react'
import { connect } from 'react-redux'
import List from '@material-ui/core/List'
import StatusesContext from './StatusesContext'
import StatusItem from './StatusItem'
import EditStatus from './EditStatus'
import DeleteStatus from './DeleteStatus'
import CreateStatus from './CreateStatus'

const SettingStatuses = props => {

  const initialState = {
    status: {
      edit: {
        id: null
      },
      delete: {
        id: null
      },
      create: {
        init: false
      }
    },
    func: {
      create: () => setState({
        ...state,
        status: {
          ...state.field,
          create: {
            init: true
          }
        }
      }),
      edit: id => setState({ 
        ...state, 
        status: { 
          ...state.field, 
          edit: { 
            id 
          } 
        } 
      }),
      delete: id => setState({
        ...state,
        status: {
          ...state.field,
          delete: {
            id
          }
        }
      }),
      default: () => setState({ ...initialState })
    }
  }

  const [ state, setState ] = useState(initialState)

  return (
    <StatusesContext.Provider value={state}>
      <List>
        {(() => {
          if (state.status.edit.id !== null)
            return (
              props.statuses.map((status, index) => 
                state.status.edit.id === status._id ? 
                  <EditStatus 
                    key={index} 
                    status={status}
                  /> : 
                  <StatusItem 
                    key={index} 
                    status={status}
                  /> 
              )
            )
          else if (state.status.delete.id !== null)
            return (
              props.statuses.map((status, index) => 
                state.status.delete.id === status._id ? 
                  <DeleteStatus 
                    key={index} 
                    status={status}
                  /> : 
                  <StatusItem 
                    key={index} 
                    status={status}  
                  /> 
              )
            )
          else if (!state.status.edit.id && !state.status.delete.id) 
            return (
              props.statuses.map((status, index) => 
                <StatusItem 
                  key={index} 
                  status={status}
                /> 
              )
            )
        })()}
        <CreateStatus />
      </List>
    </StatusesContext.Provider>
  )
}

const mapStateToProps = state => {
  return {
    statuses: state.statuses.payload
  }
}

export default connect(mapStateToProps)(SettingStatuses)