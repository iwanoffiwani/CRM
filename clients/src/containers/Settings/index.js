import React from 'react'

import Fields from './Fields/'
import Statuses from './Statuses'

import { makeStyles } from '@material-ui/core/styles'
import Container from '@material-ui/core/Container'
import Box from '@material-ui/core/Box'
import Typography from '@material-ui/core/Typography'

const useStyles = makeStyles(theme => ({
  root: {
    padding: '2em 0'
    // display: 'flex',
    // minHeight: '100vh'
  },
  box: {
    marginBottom: theme.spacing(4)
  },
  column: {
    flex: `1 0 50%`
  },
  title: {
    marginBottom: theme.spacing(4),
    fontWeight: 'bold',
    fontSize: '1.5em'
  }
}))

const Settings = props => {
  const classes = useStyles()

  return (
    <Container maxWidth='xl'>
      <div className={classes.root}>
        <Box className={classes.box}>
          <Typography 
            className={classes.title} 
            variant='h2' 
            noWrap>
            Редактирование полей
          </Typography>
          <Fields />
        </Box>
        <Box>
          <Typography 
            className={classes.title} 
            variant='h2' 
            noWrap>
            Редактирование статусов
          </Typography>
          <Statuses />
        </Box>
      </div>
    </Container>
  )
}

export default Settings