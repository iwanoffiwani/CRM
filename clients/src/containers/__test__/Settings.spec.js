import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Settings } from '../Settings'
import Box from '@material-ui/core/Box'
import Typography from '@material-ui/core/Typography'

describe('Tests Settings component', () => {
  const mapStateToProps = {
    fields: [
      {
        "value": "",
        "_id": "5dcac8a093dd3b05f8cb0adc",
        "name": "Мобильный телефон",
        "__v": 0
      },
      {
        "value": "",
        "_id": "5df9f19e1607be20381256e8",
        "name": "Номер паспорта",
        "__v": 0
      },
      {
        "value": "",
        "_id": "5dfa29e98ef6f51a186c395e",
        "name": "Домашний тел",
        "__v": 0
      },
      {
        "value": "",
        "_id": "5e00b1bfa63229263405fd21",
        "name": "Паспорт",
        "__v": 0
      }
    ],
    statuses: [
      {
        "_id": "5e28378cfa502c1edc0e7689",
        "name": "Статус-1",
        "__v": 0
      },
      {
        "_id": "5e283799fa502c1edc0e768b",
        "name": "Статус-2",
        "__v": 0
      },
      {
        "_id": "5e28379ffa502c1edc0e768c",
        "name": "Статус-4",
        "__v": 0
      },
      {
        "_id": "5e2837a5fa502c1edc0e768d",
        "name": "Статус-5",
        "__v": 0
      }
    ]
  }

  it('+++ renderer component', () => {
    const wrapper = shallow(<Settings {...mapStateToProps} />)
    expect(toJson(wrapper)).toMatchSnapshot()

    const boxWrapper = wrapper.find(Box)
    expect(boxWrapper.length).toEqual(2)

    const typographyWrapper = wrapper.find(Typography)
    expect(typographyWrapper.length).toEqual(2)
  })
})