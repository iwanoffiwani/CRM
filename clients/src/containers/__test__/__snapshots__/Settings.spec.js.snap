// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Settings component +++ renderer component 1`] = `
<Connect(Layout)>
  <div>
    <Styled(MuiBox)
      className="makeStyles-box-1"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-3"
        noWrap={true}
        variant="h2"
      >
        Редактирование полей
      </WithStyles(ForwardRef(Typography))>
      <Connect(EditFields)
        fields={
          Array [
            Object {
              "__v": 0,
              "_id": "5dcac8a093dd3b05f8cb0adc",
              "name": "Мобильный телефон",
              "value": "",
            },
            Object {
              "__v": 0,
              "_id": "5df9f19e1607be20381256e8",
              "name": "Номер паспорта",
              "value": "",
            },
            Object {
              "__v": 0,
              "_id": "5dfa29e98ef6f51a186c395e",
              "name": "Домашний тел",
              "value": "",
            },
            Object {
              "__v": 0,
              "_id": "5e00b1bfa63229263405fd21",
              "name": "Паспорт",
              "value": "",
            },
          ]
        }
      />
    </Styled(MuiBox)>
    <Styled(MuiBox)>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-3"
        noWrap={true}
        variant="h2"
      >
        Редактирование статусов
      </WithStyles(ForwardRef(Typography))>
      <Connect(EditStatuses)
        statuses={
          Array [
            Object {
              "__v": 0,
              "_id": "5e28378cfa502c1edc0e7689",
              "name": "Статус-1",
            },
            Object {
              "__v": 0,
              "_id": "5e283799fa502c1edc0e768b",
              "name": "Статус-2",
            },
            Object {
              "__v": 0,
              "_id": "5e28379ffa502c1edc0e768c",
              "name": "Статус-4",
            },
            Object {
              "__v": 0,
              "_id": "5e2837a5fa502c1edc0e768d",
              "name": "Статус-5",
            },
          ]
        }
      />
    </Styled(MuiBox)>
  </div>
</Connect(Layout)>
`;
